### PR TITLE
Add new CRDs to support root resource mapping

### PR DESCRIPTION
**Pull request by LeiQL**
- Add new CRDs:  instanceContainer, solutionContainer, targetContainer, catalogContainer, campaignContainer
- Add `rootResource` and `version` in solution, target, instance, catalog, campaign to support liquid property mapping

### CODE REVIEW

**Comment by Haishi2016** [comment]
Just to double-check, the plan is to add these CRDs first and then do renaming later?

**Comment by LeiQL** [comment]
> t to double-check, the plan is to add these CRDs first and then do renaming

@Haishi2016  renaming is not scheduled before private preview. We can add a task for renaming and schedule it based on priorities. And FYI, the support of object reference (myapp:latest, myapp:v1-0) is under development and testing. I will have PRs for the support of labelling and reference.

File: api/pkg/apis/v1alpha1/model/campaign.go [hunk 1]
```diff
@@ -125,9 +125,11 @@ func (c ActivationState) DeepEquals(other IDeepEquals) (bool, error) {
 }
 
 type CampaignSpec struct {
-	FirstStage  string               `json:"firstStage,omitempty"`
-	Stages      map[string]StageSpec `json:"stages,omitempty"`
-	SelfDriving bool                 `json:"selfDriving,omitempty"`
+	FirstStage   string               `json:"firstStage,omitempty"`
+	Stages       map[string]StageSpec `json:"stages,omitempty"`
+	SelfDriving  bool                 `json:"selfDriving,omitempty"`
+	Version      string               `json:"version,omitempty"`
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/model/campaign.go:131]: Is it better to put Version and RootResource in ObjectMeta?

File: api/pkg/apis/v1alpha1/model/campaigncontainer.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	equal, err := c.ObjectMeta.DeepEquals(otherC.ObjectMeta)
+	if err != nil || !equal {
+		return equal, err
+	}
+
+	equal, err = c.Spec.DeepEquals(*otherC.Spec)
+	if err != nil || !equal {
+		return equal, err
+	}
+
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/model/campaigncontainer.go:45]: I notice all the deepEquals doesn't compare Status not only in your PR but also in the existing code base. Do you know why?

Comment (Haishi2016) [reply api/pkg/apis/v1alpha1/model/campaigncontainer.go:45]: deepEquals is often used to detect if an object has changed. A user submitted object usually doesn't have status associated with it, while the object is retrieved from the system has status. Hence status comparison is generally skipped.

File: packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml
Comment (Haishi2016) [inline packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml:1]: should packages/helm/templates/pai-cluster-role.yaml be updated to include new object types

Comment (LeiQL) [reply packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml:1]: Version and RootResource are used to add labels. The properties are mapped using liquid template in k8s bridge.

File: api/pkg/apis/v1alpha1/model/campaign.go (hunk 1 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/model/campaign.go:131]: Version and RootResource are used to add labels. The properties are mapped using liquid template in k8s bridge.

Comment (LeiQL) [reply api/pkg/apis/v1alpha1/model/campaign.go:131]: Usage of these properties will be shown in next PR for changes in api/controllers.

File: packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml
Comment (LeiQL) [reply packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml:1]: Updated `pai-cluster-role.yaml `. Added new CRDs and `"targets/status", "solutions/status", "catalogs/status", "campaigns/status"` which are missing. Please double confirm.

````
  resources: ["*", "validatingwebhookconfigurations", "mutatingwebhookconfigurations", "rolebindings", "roles", "clusterrolebindings", "clusterroles", "secrets", "serviceaccounts", "poddisruptionbudgets", "podsecuritypolicies", "resourcequotas", "customresourcedefinitions", "targets", "skills", "models", "skillpackages", "sites/status", "activations/status", "campaigns", "activations", "sites", "catalogs", "devices", "instances", "solutions", "deployments", "services", "devices/status", "instances/status", "targets/status", "solutions/status", "catalogs/status", "campaigns/status", "namespaces", "targetcontainers", "solutioncontainers", "instancecontainers", "catalogcontainers", "campaigncontainers", "targetcontainers/status", "solutioncontainers/status", "instancecontainers/status", "catalogcontainers/status", "campaigncontainers/status"]

**Review by msftcoderdjw — COMMENTED** [review]
This PR introduces the first part of versioning changes - modify data model. I am OK with this change. Team can review and drop more comments if needed. One more suggestion is maybe we don't need to keep rootResource and version in Spec. It is better to keep them in metadata like metadata.name.

**Comment by LeiQL** [comment]
> This PR introduces the first part of versioning changes - modify data model. I am OK with this change. Team can review and drop more comments if needed. One more suggestion is maybe we don't need to keep rootResource and version in Spec. It is better to keep them in metadata like metadata.name.

**Comment by LeiQL** [comment]
@Haishi2016 @msftcoderdjw do we have any other concerns or questions on this PR?

**Comment by msftcoderdjw** [comment]
@LeiQL please rebase your PR.

**Review by msftcoderdjw — APPROVED** [review]

**Pull request merged**